Windows version check in doctor

eliasyishak opened the pull request:
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.


Addresses #102607 stating that a warning should be emitted during a `flutter doctor` run to warn users on Windows versions less than 10

eliasyishak committed:
validator created -- logic to be developed on windows

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
logic added for validator -- to be tested in windows env next

eliasyishak committed:
making call to built in windows 'systeminfo' from dart

eliasyishak committed:
Update windows_version_validator.dart

eliasyishak committed:
conditional to only run on windows

eliasyishak committed:
Update doctor.dart

eliasyishak committed:
Merge pull request #1 from eliasyishak/master

merging version check

eliasyishak committed:
clean up

eliasyishak committed:
conditional update to check for version

eliasyishak committed:
conditional status info message added

eliasyishak commented:
Example return from a `systeminfo` call in a Windows environment.

```
Host Name:                 XXXXXXXXXXXX
OS Name:                   Microsoft Windows 10 Enterprise
OS Version:                10.0.19044 N/A Build 19044
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Member Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          N/A
Registered Organization:   N/A
Product ID:                XXXXXXXXXXXX
Original Install Date:     8/4/2022, 2:51:28 PM
System Boot Time:          8/10/2022, 1:03:10 PM
System Manufacturer:       Google
System Model:              Google Compute Engine
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: AMD64 Family 23 Model 49 Stepping 0 AuthenticAMD ~2250 Mhz
BIOS Version:              Google Google, 6/29/2022
Windows Directory:         C:\\Windows
System Directory:          C:\\Windows\\system32
Boot Device:               \\Device\\HarddiskVolume2
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-08:00) Pacific Time (US & Canada)
Total Physical Memory:     32,764 MB
Available Physical Memory: 17,852 MB
Virtual Memory: Max Size:  33,788 MB
Virtual Memory: Available: 18,063 MB
Virtual Memory: In Use:    15,725 MB
Page File Location(s):     C:\\pagefile.sys
Domain:                    ad.corp.google.com
Logon Server:              \\CBF-DC-8
Hotfix(s):                 7 Hotfix(s) Installed.
                           [01]: KB5013624
                           [02]: KB5003791
                           [03]: KB5012170
                           [04]: KB5016616
                           [05]: KB5014032
                           [06]: KB5014671
                           [07]: KB5015895
Hyper-V Requirements:      A hypervisor has been detected. Features required for Hyper-V will not be displayed.
```

Each line will be parsed by the windows version validator by filtering for the line that is relevant.

eliasyishak commented:
cc @Jasguerrero

Jasguerrero commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,88 @@
+  @override
+  Future<ValidationResult> validate() async {
+    final ProcessResult result = await Process.run('systeminfo', <String>[]);
+    final String resultStdout = result.stdout as String;
```
You need to check if `ProcessResult.exitCode` was successful before going further

eliasyishak replied:
Oh good point, I am adding a try, catch to find any general ProcessExceptions now and then also doing a check afterwords on the exit code itself

Jasguerrero commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,88 @@
+      version = null;
+    }
+
+    ValidationType windowsVersionStatus = ValidationType.missing;
```
This can be final and set it to `ValidationType.missing` on the else

Jasguerrero commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,88 @@
+    bool versionText = false;
+    bool versionSemver = false;
+    String? version;
+    for (int i = 0; i < systemInfoElements.length; i++) {
```
nit: you can iterate directly over the `systemInfoElements` (as well as the `lineElements`) and avoid using `.elementAt`

```
for (final String curLine in systemInfoElements) {...}
```

eliasyishak replied:
nice, i meant to actually refactor this to use that notation

eliasyishak committed:
refactoring to not use `dart:io`

eliasyishak committed:
changes requested during review

eliasyishak committed:
adding documentation for validator

eliasyishak committed:
fix to license file

eliasyishak committed:
fix to license file

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge branch 'windows-version-check-in-doctor' of https://github.com/eliasyishak/flutter into windows-version-check-in-doctor

eliasyishak committed:
Merge branch 'windows-version-check-in-doctor' into master

eliasyishak committed:
Merge pull request #2 from eliasyishak/master

merging for latest code

eliasyishak committed:
fixing merging error

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,117 @@
+import '../doctor_validator.dart';
+
+class WindowsVersionValidator extends DoctorValidator {
+  /// Validator to be run with `flutter doctor` to check
+  /// Windows host machines if they are running supported versions,
+  /// current unsupported versions = 7, 8
```
Nit, add documentation comments like these to the class declaration and not the constructor.

Follow https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary guidelines for structuring the comment.

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,117 @@
+  Future<ValidationResult> validate() async {
+    final ProcessResult result;
+    try {
+      result = _processManager.runSync(<String>['systeminfo']);
```
rather than use runSync, use run and await the result. Since many doctor validators run concurrently and runSync will block on output.

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,117 @@
+      return const ValidationResult(
+        ValidationType.missing,
+        <ValidationMessage>[],
+        statusInfo:
+            'Unable to run Windows version check using built-in `systeminfo`',
```
Flutter developers are unlikely to know what this means. I would loop in @christopherfujino and @Jasguerrero , but I would epect that if this check fails because we can't determine the version, we should just not say anything at all rather than worry people.

eliasyishak replied:
Agreed, I wasn't too sure how to handle errors/failures for the validator... i just added them to handle all errors during testing

christopherfujino replied:
I think for both this and exitCode != 0, we have two options:

1) show what happened to the user and ask them to file a bug because something we weren't expecting happened; or
2) do what Jonah suggested, and not say anything and not worry people (anytime there's a non-green doctor validator it causes a subset of our users a lot of anxiety)

If you want to do 1, you can just not catch the ProcessException and then explicitly throw an exception for the exitCode != 0 case; uncaught exceptions from doctor validators will be caught higher up and surfaced to the user.

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,117 @@
+      );
+    }
+
+    if (result.exitCode != 0) {
```
Same here

christopherfujino replied:
Do you know in what cases `systeminfo` will return non-zero?

eliasyishak replied:
I can't find any examples of why this would return non-zero but just wanted to cover all basis because the following code would rely on a successful exit code

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,117 @@
+    bool versionText = false;
+    bool versionSemver = false;
+    String? version;
+    for (final String curLine in systemInfoElements) {
```
Rather than breaking down each line into individual chunks and running the regex over that, use something like String.contains on each line and pull the version info out of the match object returned.

eliasyishak replied:
That's a good point, I'm wondering if it would be worth ditching looping over each line and use the `multiLine` boolean on `RegExp` for matches. What are your thoughts on that approach?

eliasyishak replied:
This link should show the pattern working in a demo.

https://regex101.com/r/EA9AAq/1

Using the above demo, the middle group would be our version we're interested in

jonahwilliams replied:
I would not use multiLine unless you potentially want a match that spans multiple lines

jonahwilliams replied:
If you have something that matches your line, you can just run that over the whole string and it should work

eliasyishak replied:
sounds good, editing the pattern matcher to account for that

eliasyishak committed:
added 3 tests for each return case for windows version

- case 1: successful check (version 10 or greater)
- case 2: `systeminfo` process call fails
- case 3: version identified is less than 10

eliasyishak committed:
update to documentation + using async process run

eliasyishak commented on packages/flutter_tools/lib/src/doctor.dart:
```diff
@@ -129,6 +130,9 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
         flutterRoot: () => Cache.flutterRoot!,
         operatingSystemUtils: globals.os,
       ),
+      if (platform.isWindows) WindowsVersionValidator(
```
I'm not aware of any versioning constraints on the other platforms, i would be more than happy to chase down an answer if you can point me to someone i can chat – after that i can look into making a more general OS validator

christopherfujino replied:
actually, it looks like we don't have a minimum macOS version, so I guess we're good for now. Your macOS version will be transitively constrained by our supported xcode version, but we already have a validator for that.

eliasyishak committed:
simplified string parsing to utilize builtin regex functionality

eliasyishak committed:
remove todo

christopherfujino commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,77 @@
+    // Regular expression pattern for identifying
+    // semantic versioned strings
+    // (ie. 10.5.4123)
+    final RegExp regex =
```
I would pull this into a @visibleForTesting method that takes the stdout string and then returns `Iterable<RegExpMatch> matches`, and then you can unit test different strings

eliasyishak committed:
add optional indicator for new line chars

christopherfujino commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -50,7 +50,7 @@ class WindowsVersionValidator extends DoctorValidator {
     // (ie. 10.5.4123)
     final RegExp regex =
-        RegExp(r'\n(OS Version:\s*)([0-9]+\.[0-9]+\.[0-9]+)(.*)\n');
+        RegExp(r'\n?(OS Version:\s*)([0-9]+\.[0-9]+\.[0-9]+)(.*)\n?');
```
Does adding the question mark not make this also match `'BIOS Version: `?

eliasyishak replied:
I don't believe so, the question mark should be only looking for 0 or 1 newline characters before `OS`. Unless I had the wrong assumption there?

I added the question mark incase there was no new line character before the `OS Version` line on print out

christopherfujino replied:
So in "BIOS Version:" won't there be 0 newline characters before "OS"?

christopherfujino replied:
whether i'm right or wrong, this should be test cases in the unit tests :)

eliasyishak replied:
I see what you mean now, I'm going to try a refactor with `multiLine` enabled so that I can specify the start of the line as a reference point

eliasyishak replied:
I'll build some test cases with multiline enabled when i get back. thanks for the review! 😁

eliasyishak committed:
fixing regex pattern to look for start of line

zanderso commented:
Oops. I fatfingered the ready for review button on mobile, but there's no way in the mobile UI to change back to draft...

jonahwilliams commented:
Switch to desktop view 😁

eliasyishak committed:
added unit test for regex pattern validator

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge pull request #3 from eliasyishak/windows-version-check-in-doctor

Windows version check in doctor

eliasyishak committed:
Merge pull request #4 from eliasyishak/master

updating feature branch with upstream master

eliasyishak committed:
removing ProcessException catch since doctor validator handles it already

jonahwilliams commented on packages/flutter_tools/lib/src/doctor.dart:
```diff
@@ -129,6 +130,9 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
         flutterRoot: () => Cache.flutterRoot!,
         operatingSystemUtils: globals.os,
       ),
+      if (platform.isWindows) WindowsVersionValidator(
+        processManager: globals.processManager,
+      ),
```
```suggestion
      if (platform.isWindows)
        WindowsVersionValidator(
          processManager: globals.processManager,
        ),
```

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,79 @@
+import '../base/io.dart';
+import '../doctor_validator.dart';
+
+const List<String> unsupportedVersions = <String>[
```
```suggestion
/// Flutter only supports development on Windows host machines version 10 and greater.
const List<String> kUnsupportedVersions = <String>[
```

Or something like that

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,79 @@
+  '8',
+];
+
+/// Validator to be run with `flutter doctor` to check
+/// Windows host machines if they are running supported versions.
+class WindowsVersionValidator extends DoctorValidator {
```
```suggestion
/// Validator for supported Windows host machine operating system version.
class WindowsVersionValidator extends DoctorValidator {
```

Since all validators are used by doctor

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,79 @@
+  /// Provide a literal string as the Regex pattern
+  /// and a string to validate and get a boolean determining
+  /// if the string has at least one match
+  static Iterable<RegExpMatch> validateString(String pattern, String str,
+      {bool multiLine = true}) {
+    final RegExp regex = RegExp(
+      pattern,
```
I don't really think this helper method is carrying its weight. First, I would move the regex string to a top level and document there what it is intended to match. Then you can replace it with `regex.allMaches(contents)`

jonahwilliams replied:
Oh, I think from reading @christopherfujino's comment, the idea was to embed the version matching pattern in this method to make it easy to test. 

From reading the test cases, I think it would work OK without this as a public method. But if you're not going to specifically test it you should drop this.

eliasyishak replied:
Yeah, Chris mentioned that it would be helpful to make this regex helper a method that we can unit test independently, so that we can feed it test strings as we start to encounter newer versions of the `stdout` from `systeminfo`.

jonahwilliams replied:
You're basically not testing this method, though, because this method itself is trivial. I would just test the regex itself

christopherfujino replied:
I was thinking that this helper method would also include lines 58-61 and return a boolean. Also the rather than passing the pattern (which is actually what you want to test), that the function would receive `resultStdout`.

Then your test can look (something) like:

```
test('validateVersionString() correctly validates the output of `systeminfo`', () {
  bool result = validateVersionString('123');
  expect(result, isFalse);
  result = validateVersionString('OS Version:   8.0.0');
  // Too old
  expect(result, isFalse);
  result = validateVersionString('OS Version:     11.1.2');
  expect(result, isTrue);
});
```

jonahwilliams replied:
I think this approach is fine too, its the middle ground that isn't very useful

eliasyishak replied:
I do agree with @jonahwilliams that the static method was very trivial but @christopherfujino what your test looks like is essentially what the entire validator was built to check.

I shied away from creating a regex method that returned a boolean because I didn't want to assume that there would only be one match within the `stdout`. The conditionals beginning on line 60 were added to ensure that the matches only returned 1 result AND that the major version was valid.

christopherfujino replied:
ok, makes sense, i'm fine with you removing this function

jonahwilliams commented on packages/flutter_tools/test/general.shard/windows_version_validator_test.dart:
```diff
@@ -0,0 +1,209 @@
+  statusInfo: 'Unable to confirm if installed Windows version is 10 or greater',
+);
+
+void main() {
```
I would add a test for non-zero exit code.

jonahwilliams commented on packages/flutter_tools/lib/src/windows/windows_version_validator.dart:
```diff
@@ -0,0 +1,79 @@
+    // Use the string split method to extract the major version
+    // and check against the [unsupportedVersions] list
+    final ValidationType windowsVersionStatus;
+    String statusInfo;
```
```suggestion
    final String statusInfo;
```

eliasyishak committed:
Apply Jonah suggestions from code review

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

eliasyishak committed:
renaming variable

eliasyishak committed:
additional test added for a nonzero exit code from systeminfo

eliasyishak committed:
removing trivial static method, added constant for semver pattern

jonahwilliams reviewed (approved):
LGTM
